spock-specs/src/test/groovy/org/spockframework/smoke/ast/AstSpec.groovy

```diff
@@ -0,0 +1,211 @@
+    this.getSpecificationContext().getMockController().leaveScope()
+}'''
+  }
+
+
+  def "astToSourceFeatureBody can render everything"() {
+    when:
+    def result = compiler.astToSourceFeatureBody('''
+    given:
+    def nothing = null
+    ''', Show.all(), CompilePhase.INSTRUCTION_SELECTION)
+
+    then:
+    result == '''\
+package apackage
+
+import spock.lang.*
+
+@org.spockframework.runtime.model.SpecMetadata(filename = 'scriptXXXXX.groovy', line = 1)
+public class apackage.ASpec extends spock.lang.Specification {
+
+    @org.spockframework.runtime.model.FeatureMetadata(name = 'a feature', ordinal = 0, line = 1, blocks = [org.spockframework.runtime.model.BlockKind.SETUP[]], parameterNames = [])
+    public void $spock_feature_0_0() {
+        java.lang.Object nothing = null
+        this.getSpecificationContext().getMockController().leaveScope()
+    }
+
+}'''
+  }
+
+  def "full feature exercise"() {
+    when:
+    def result = compiler.astToSource('''
+    package apackage
+
+    import groovy.transform.EqualsAndHashCode
+
+    abstract class Foo implements Comparable {
+      List<String> x = new ArrayList<>()
+
+
+    }
+
+    @EqualsAndHashCode
+    class Bar extends Foo {
+      static final int[] ARR = [1, 2, 3] as int[]
+      static final String STR = 'str'
+      static final char CHR = ' '
+      static final int INT = 10
+      static int st
+      int order
+
+      Bar() {
+       println '42'
+      }
+
+      static {
+        st = 42
+      }
+
+      {
+        order = 1
+      }
+
+      void method() {
+        outer:
+        for (int i = 0; i < INT; i++) {
+          if(i % 2 == 0) {
+            print 'even'
+            continue outer
+          } else {
+            print 'odd'
+          }
+          break outer
+        }
+        for(String y : x) {
+          print(y)
+        }
+      }
+
+      @Override
+      int compareTo(Object o){
+        return order <=> (o as Bar).order
+      }
+    }
+
+    class Ext <T extends Serializable> {
+      def foo (List<? super T> consumer){}
+    }
+    ''')
+
+    then:
+    result == '''\
+package apackage
+
+import groovy.transform.EqualsAndHashCode as EqualsAndHashCode
+
+public abstract class apackage.Foo extends java.lang.Object implements java.lang.Comparable {
+
+    private java.util.List<String> x
+
+}
+package apackage
+
+import groovy.transform.EqualsAndHashCode as EqualsAndHashCode
+
+@groovy.transform.EqualsAndHashCode
+public class apackage.Bar extends apackage.Foo {
+
+    private static final [I ARR
+    private static final java.lang.String STR = 'str\'
+    private static final char CHR = ' \'
+    private static final int INT = 10
+    private static int st
+    private int order
+
+    public apackage.Bar() {
+        this.println('42')
+    }
+
+    {
+        order = 1
+    }
+
+    static {
+        st = 42
+    }
+
+    public void method() {
+        outer:
+        for (java.lang.Integer i = 0; i < INT ;( i )++) {
```

Comment:
Can we leave out the parens around the `i` if it is a simple variable expression?

Reply:
It would be possible, but it would increase the complexity by introducing special case handling. 
IMHO the goal is to get a meaningful representation of the code, while keeping the renderer simple enough to maintain.

In any case I would see that as icing on top, and would defer it to a later change.

Reply:
yeah, you are probably right. Should be enough if it were correct, does not have to be pretty.
A pitty that it is not really correct what is generated, the Groovy guys should put more love into this converter, they should know better what they are doing. :-D